### PR TITLE
Stripe components

### LINKED ISSUE #1462: [ACTION] Stripe new actions
**Is there a specific app this action is for?**
Stripe

**Please provide a link to the relevant API docs for the specific service / operation.**
https://stripe.com/docs/api

Payment Intents ([guide](https://stripe.com/docs/payments/payment-intents), [reference](https://stripe.com/docs/api/payment_intents))

- [x] **List Payment Intents** | Retrieves a list of payment intents or filters by customer ID
- [x] **Create a Payment Intent** | Creates a payment intent
- [x] **Retrieve a Payment Intent** | Retrieves the details of a payment intent that has previously been created
- [x] **Confirm a Payment Intent** | Confirms a payment intent by its ID. Upon confirmation, the payment intent will attempt to initiate a payment
- [x] **Capture a Payment Intent** | Captures an existing uncaptured payment intent by its ID
- [x] **Update a Payment Intent** | Updates the specific payment intent by setting the values of the parameters passed
- [x] **Cancel a Payment Intent** | Cancels a payment intent by its ID. For payment intents with status 'requires capture' the remaining capturable amount will automatically be refunded

[Customers](https://stripe.com/docs/api/customers)

- [x] **List Customers** | Retrieves a list of customers filtered by criteria
- [x] **Create a Customer** | Creates a new customer object
- [x] **Retrieve a Customer** | Retrieves the details of an existing customer
- [x] **Update a Customer** | Updates the specified customer by setting the values of the parameters passed
- [x] **Delete a Customer** | Permanently delete a customer. It cannot be undone

Refunds ([guide](https://stripe.com/docs/refunds), [reference](https://stripe.com/docs/api/refunds))

- [x] **List All Refunds** | Returns a list of all refunds you've previously created
- [x] **Create a Refund** | Creates a new refund
- [x] **Retrieve a Refund** | Retrieves the details of an existing refund
- [x] **Update a Refund** | Updates the specified refund by setting the values of the parameters passed

Payouts ([guide](https://stripe.com/docs/payouts), [reference](https://stripe.com/docs/api/payouts))

- [x] **List All Payouts** | Returns a list of existing payouts sent to third-party bank accounts or that Stripe has sent you
- [x] **Create a Payout** | To send funds to your own bank account, you create a new payout object
- [x] **Cancel a Payout** | Cancels a previously created payout
- [x] **Retrieve a Payout** | Retrieves the details of an existing payout
- [x] **Update a Payout** | Updates the specific payout by setting the values of the parameters passed

Balance ([guide](https://stripe.com/docs/connect/account-balances), [reference](https://stripe.com/docs/api/balance))

- [x] **Retrieve Balance** | Retrieves the current account balance
- [x] **List All Balance History** | Returns a list of transactions that have contributed to the Stripe account balance
- [ ] ~~**Retrieve a Balance Transaction** | Retrieves the balance transaction with the given ID~~

Invoices ([guide](https://stripe.com/docs/invoicing/sending), [reference](https://stripe.com/docs/api/invoices))

- [x] **Search Invoices** | Retrieves a list of invoices filtered by criteria
- [ ] ~~**List All Invoice Line Items** | Retrieves a list of invoice line items by the invoice ID~~
- [x] **Create an Invoice** | Creates a new invoice by the customer ID
- [x] **Retrieve an Invoice** | Retrieves the details of an existing invoice with the given ID
- [x] **Update an Invoice** | Updates the specified invoice by setting the values of the parameters passed
- [x] **Finalize a Draft Invoice** | Finalizes a draft invoice by its ID
- [x] **Void or Delete an Invoice** | Permanently deletes a draft invoice, or voids a finalized or subscription invoice with the given ID. This cannot be undone
- [x] **Write off Invoice** | Write off an uncollectible invoice
- [ ] ~~**Search Invoice Items** | Retrieves a list of invoice items filtered by criteria~~
- [x] **Create an Invoice Item** | Creates a new invoice item to be added to a draft invoice
- [x] **Retrieve and Invoice Item** | Retrieves the details of an existing invoice item with a given ID
- [x] **Update an Invoice Item** | Updates the specified invoice item by setting the values of the parameters passed
- [x] **Delete an Invoice Item** | Permanently delete an invoice item which is not attached to invoices, or if it's attached to a draft invoice

Checkout Sessions ([guide](https://stripe.com/docs/payments/checkout), [reference](https://stripe.com/docs/api/checkout/sessions))

- [x] **Retrieve a Checkout Session** | Retrieve a details of a checkout session and all of it's line items with the given ID

Prices ([guide](https://stripe.com/docs/billing/prices-guide), [reference](https://stripe.com/docs/api/prices))

- [x] **Retrieve a Price** | Retrieves the details of a price with the given ID

Metered Billing ([guide](https://stripe.com/docs/billing/subscriptions/metered-billing), [reference](https://stripe.com/docs/api/usage_records))

- [x] **Create a Usage Record** | Creates a usage record for a specified subscription item and date, and fills it with a quantity

### PR DESCRIPTION
**Pull request by compwright**
Resolves #1462

### CODE REVIEW

File: components/stripe/actions/cancel-or-reverse-payout/cancel-or-reverse-payout.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "A payout can be canceled only if it has not yet been paid out. A payout can be reversed " +
+    "only if it has already been paid out. Funds will be refunded to your available balance.",
+  props: {
+    stripe,
+    id: {
+      propDefinition: [
+        stripe,
+        "payout",
+      ],
+      required: true,
```
Comment (dylburger) [inline components/stripe/actions/cancel-or-reverse-payout/cancel-or-reverse-payout.js:17]: This is the default, so you can remove and the property is technically `optional` (`false` by default): https://pipedream.com/docs/components/api/#user-input-props

File: components/stripe/actions/cancel-payment-intent/cancel-payment-intent.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  props: {
+    stripe,
+    id: {
+      propDefinition: [
+        stripe,
+        "payment_intent",
+      ],
+      required: true,
+    },
+    cancellation_reason: {
```
Comment (dylburger) [inline components/stripe/actions/cancel-payment-intent/cancel-payment-intent.js:21]: Is this optional? If so you'll want to add `optional: true` either here or in the prop definition in the app file.

Comment (dylburger) [reply components/stripe/actions/cancel-payment-intent/cancel-payment-intent.js:21]: The UI doesn't yet communicate prop optionality well, and that's on the backlog.

File: components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  props: {
+    stripe,
+    id: {
+      propDefinition: [
+        stripe,
+        "payment_intent",
+      ],
+      required: true,
+    },
+    customer: {
```
Comment (dylburger) [inline components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js:19]: Is this used?

File: components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        "customer",
+      ],
+    },
+    payment_method: {
+      propDefinition: [
+        stripe,
+        "payment_method",
+      ],
+    },
+    receipt_email: {
```
Comment (dylburger) [inline components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js:31]: Is this used?

File: components/stripe/actions/capture-payment-intent/capture-payment-intent.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+const pick = require("lodash.pick");
+const stripe = require("../../stripe.app.js");
+
+module.exports = {
+  key: "stripe-capture-payment-intent",
+  name: "Capture a Payment Intent",
+  version: "0.0.1",
+  description: "Capture the funds of an existing uncaptured payment intent.",
```
Comment (dylburger) [inline components/stripe/actions/capture-payment-intent/capture-payment-intent.js:8]: Very small nit, but I typically don't include a period at the end of descriptions that aren't full sentences

File: components/stripe/actions/create-payment-intent/create-payment-intent.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      ],
+      required: true,
+    },
+    country: {
+      propDefinition: [
+        stripe,
+        "country",
+      ],
+      required: true,
+      default: "us",
```
Comment (dylburger) [inline components/stripe/actions/create-payment-intent/create-payment-intent.js:24]: Does it make sense to incorporate defaults like this into the prop definition itself?

File: components/stripe/stripe.app.js [hunk 7]
```diff
@@ -1,13 +1,394 @@
-const stripe = require("stripe")
+const stripe = require("stripe");
+
+const createOptionsMethod = (collectionOrFn, keysOrFn) => async function ({ prevContext }) {
```
Comment (dylburger) [inline components/stripe/stripe.app.js:3]: Nice

**Review by dylburger — CHANGES_REQUESTED** [review]
Small feedback, mostly around props:

- Props are required by default, and the property is technically `optional` (which is `false` by default, so all props are required unless specified otherwise): https://pipedream.com/docs/components/api/#user-input-props . So you'll need to swap the logic for these props, removing `required: true` and setting `optional` to `true` where relevant.
- I noticed a case (e.g. with the `confirm-payment-intent` action) where it didn't look like we were passing specific props to the final method. You may want to go through and confirm there aren't other cases I missed.
- Other small comments inline

File: components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js (hunk 3 above)
Comment (compwright) [reply components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js:20]: @dylburger good catch, and yes, but not in an obvious place, so I added an explanatory comment.

File: components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js (hunk 4 above)
Comment (compwright) [reply components/stripe/actions/confirm-payment-intent/confirm-payment-intent.js:32]: @dylburger missed that one! thanks :)

**Comment by compwright** [comment]
@dylburger thanks a lot, that was very helpful feedback! I ended up making all props optional by default, then turning off optionality on the required ones. I personally like annotating all required fields better than annotating all optional fields, but that's just me. Usually there are fewer required fields than optional ones, which is certainly true here.

**Comment by dylburger** [comment]
Yes, we've gone back and forth about that internally, too. Stripe is definitely a good example of an API that provides extensive optional properties.

**Review by dylburger — CHANGES_REQUESTED** [review]
I only caught this as I was testing and I forget this all the time myself, but you'll need to add

```
type: "action"
```

to all components, as well, to inform the product that these are actions. 

We don't have a `type: "source"` yet (components that don't have a type are assumed to be sources), but we plan to add that and throw errors on `pd publish` to make that more clear.

Testing a couple more actions now, but if you don't hear any other comments from me, assume these are good to merge once you dd the `type` property.

File: components/stripe/actions/delete-customer/delete-customer.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  key: "stripe-delete-customer",
+  name: "Delete a Customer",
+  version: "0.0.1",
+  description: "Delete a customer",
+  props: {
+    stripe,
+    customer: {
+      propDefinition: [
+        stripe,
+        "name",
```
Comment (dylburger) [inline components/stripe/actions/delete-customer/delete-customer.js:13]: Shouldn't this take the customer ID? https://stripe.com/docs/api/customers/delete

**Comment by compwright** [comment]
@dylburger thanks for the review. Updated eslint rules are passing, but I had to add some overrides. Please review https://github.com/PipedreamHQ/pipedream/pull/1552/commits/c9bd4bfe0b4efa0da831481fe4bfe759767fd8f8

**Pull request merged**